rcl/test/rcl/test_publisher.cpp

```diff
@@ -490,3 +535,294 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_invalid_publish
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_publish_loaned_message, [](auto...) {return RMW_RET_ERROR;});
+    EXPECT_EQ(RCL_RET_ERROR, rcl_publish_loaned_message(&publisher, &msg, nullptr));
+  }
+  {
+    // mocked, borrow loaned nominal usage
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_borrow_loaned_message, [](auto...) {return RMW_RET_OK;});
+    EXPECT_EQ(RCL_RET_OK, rcl_borrow_loaned_message(&publisher, ts, &msg_pointer));
+  }
+  {
+    // bad params borrow loaned
+    EXPECT_EQ(RCL_RET_PUBLISHER_INVALID, rcl_borrow_loaned_message(nullptr, ts, &msg_pointer));
+    EXPECT_EQ(
+      RCL_RET_PUBLISHER_INVALID, rcl_borrow_loaned_message(&not_init_publisher, ts, &msg_pointer));
+  }
+  {
+    // mocked, nominal return loaned message
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_return_loaned_message_from_publisher, [](auto...) {return RMW_RET_OK;});
+    EXPECT_EQ(RCL_RET_OK, rcl_return_loaned_message_from_publisher(&publisher, &msg));
+  }
+  {
+    // bad params return loaned message
+    EXPECT_EQ(
+      RCL_RET_PUBLISHER_INVALID,
+      rcl_return_loaned_message_from_publisher(nullptr, &msg));
+    EXPECT_EQ(
+      RCL_RET_PUBLISHER_INVALID,
+      rcl_return_loaned_message_from_publisher(&not_init_publisher, &msg));
+    EXPECT_EQ(
+      RCL_RET_INVALID_ARGUMENT,
+      rcl_return_loaned_message_from_publisher(&publisher, nullptr));
+  }
+
+  test_msgs__msg__BasicTypes__fini(&msg);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_mocks_fail_init) {
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  rcl_ret_t ret = RCL_RET_OK;
+
+  {
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_validate_node_name, [](auto...) {return RMW_RET_ERROR;});
+    ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+    rcl_reset_error();
+  }
+  {
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_validate_node_name, [](auto...) {return RMW_RET_INVALID_ARGUMENT;});
+    ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+    rcl_reset_error();
+  }
+  {
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rcutils_string_map_fini, [](auto...) {return RCUTILS_RET_ERROR;});
+    ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+    rcl_reset_error();
+  }
+  {
+    auto mock = mocking_utils::patch(
```

Comment:
@Blast545 meta, unrelated: this one looks odd and a candidate for @brawner's fault injection tooling.